doc/developer-guide/internal-libraries/MemArena.en.rst

```diff
@@ -26,211 +25,136 @@
-
-After the next :func:`MemArena::alloc`:
-
-.. uml::
-   :align: center
-
-   component [block\nnew generation] as b3
-   component [block] as b4
-   component [block] as b5
-   component [block] as b6
-   component [block] as b7
-
-
-   b3 -> b4
-   b4 -> b5
-   b5 -> b6
-   b6 -> b7
-
-   generation -u- b3
-   current -u- b3
-
-A caller can actually :func:`MemArena::alloc` **any** number of bytes. Internally, if the arena is unable to allocate enough memory for the allocation, it will create a new internal block of memory large enough and allocate from that. So if the arena is allocated like:
-
-.. code-block:: cpp
-   
-   ts::MemArena *arena = new ts::MemArena(64);
-
-The caller can actually allocate more than 64 bytes. 
-
-.. code-block:: cpp
-
-   ts::MemSpan span1 = arena->alloc(16);
-   ts::MemSpan span1 = arena->alloc(256);
-
-Now, span1 and span2 are in the same generation and can both be safely used. After:
-
-.. code-block:: cpp
-
-   arena->freeze();
-   ts::MemSpan span3 = arena->alloc(512);
-   arena->thaw();
-
-span3 can still be used but span1 and span2 have been deallocated and usage is undefined. 
-
-Internal blocks are adjusted for optimization. Each :class:`MemArena::Block` is just a header for the underlying memory it manages. The header and memory are allocated together for locality such that each :class:`MemArena::Block` is immediately followed with the memory it manages. If a :class:`MemArena::Block` is larger than a page (defaulted at 4KB), it is aligned to a power of two. The actual memory that a :class:`MemArena::Block` can allocate out is slightly smaller. This is because a portion of the allocated memory is reserved for the header. Another 16 bytes is reserved to track the allocation headers used by malloc; for page alignment. ie, the default block size is 32768 bytes, but it will only be able to allocate out 32720 bytes. 
+When a |MemArena| instance is constructed no memory is reserved. A hint can be provided so that the
+first internal reservation of memory will have close to but at least that amount of free space
+available to be allocated.
+
+In normal use memory is allocated from |MemArena| using :func:`MemArena::alloc` to get chunks
+of memory, or :func:`MemArena::make` to get constructed class instances. :func:`MemArena::make`
+takes an arbitrary set of arguments which it attempts to pass to a constructor for the type
+:code:`T` after allocating memory (:code:`sizeof(T)` bytes) for the object. If there isn't enough
+free reserved memory, a new internal block is reserved. The size of the new reserved memory will be at least
+the size of the currently reserved memory, making each reservation larger than the last.
+
+The arena can be **frozen** using :func:`MemArena::freeze` which locks down the currently reserved
+memory and forces the internal reservation of memory for the next allocation. By default this
+internal reservation will be the size of the frozen allocated memory. If this isn't the best value a
```

Comment:
A comment in another [thread](https://github.com/apache/trafficserver/pull/3901#discussion_r205848561) has implications this statement: `By default this internal reservation will be the size of the frozen allocated memory.`

Leaving this comment here since the other thread is hidden now in my browser.

Reply:
I left a comment elsewhere, but this is true because the allocated memory size is tracked. The number of reserved blocks used is irrelevant.

Reply:
Looks good.